src/eos-update-server.c

```diff
@@ -0,0 +1,863 @@
+    { NULL }
+  };
+
+  context = g_option_context_new ("Endless OSTree server");
+  group = g_option_group_new (NULL, NULL, NULL, options, NULL);
+  g_option_group_add_entries (group, entries);
+  g_option_context_set_main_group (context, g_steal_pointer (&group));
+
+  memset (options, 0, sizeof (*options));
+  options->timeout_seconds = 5;
+  options->served_remote = g_strdup ("eos");
+
+  return g_option_context_parse (context, argc, argv, error);
+};
+
+static void
+options_clear (Options *options)
+{
+  if (options == NULL)
+    return;
+
+  options->local_port = 0;
+  options->timeout_seconds = 0;
+  g_clear_pointer (&options->served_remote, g_free);
+}
+
+G_DEFINE_AUTO_CLEANUP_CLEAR_FUNC (Options, options_clear)
+
+static void
+clear_source (guint *id)
+{
+  guint source_id;
+
+  if (id == NULL || *id == 0)
+    return;
+  source_id = *id;
+  *id = 0;
+
+  g_source_remove (source_id);
+}
+
+typedef struct Data Data;
+
+struct Data
+{
+  OstreeRepo *repo;
+  GMainLoop *loop;
+  SoupServer *server;
+
+  gint timeout_seconds;
+  guint timeout_id;
+
+  gchar *remote_name;
+
+  gchar *cached_repo_root;
+  GBytes *cached_config;
+
+  guint async_requests_pending;
+};
+
+#define DATA_CLEARED { NULL, NULL, NULL, 0, 0u, NULL, NULL, NULL, 0 }
+
+static void
+data_setup_timeout (Data *data);
+
+static gboolean
+timeout_cb (gpointer data_ptr)
+{
+  Data *data = data_ptr;
+
+  if (data->async_requests_pending > 0)
+    {
+      message ("%u asynchronous requests pending, resetting timeout", data->async_requests_pending);
+      data_setup_timeout (data);
+    }
+  else
+    {
+      message ("Timeout passed, quitting");
+      g_main_loop_quit (data->loop);
+      data->timeout_id = 0;
+    }
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+data_setup_timeout (Data *data)
+{
+  clear_source (&data->timeout_id);
+  if (data->timeout_seconds > 0)
+    data->timeout_id = g_timeout_add_seconds (data->timeout_seconds, timeout_cb, data);
+}
+
+static GKeyFile *
+key_file_copy (GKeyFile *key_file,
+               GError **error)
+{
+  g_autoptr(GKeyFile) dup = NULL;
+  gsize len = 0;
+  g_autofree gchar *raw = NULL;
+
+  raw = g_key_file_to_data (key_file, &len, error);
+  if (raw == NULL)
+    return NULL;
+
+  dup = g_key_file_new ();
+  if (!g_key_file_load_from_data (key_file,
+                                  raw,
+                                  len,
+                                  G_KEY_FILE_NONE,
+                                  error))
+    return NULL;
+
+  return g_steal_pointer (&dup);
+}
+
+static gboolean
+data_generate_config (Data *data,
+                      GError **error)
+{
+  g_autoptr(GKeyFile) config = key_file_copy (ostree_repo_get_config (data->repo),
+                                              error);
+  g_autofree gchar *raw = NULL;
+  gsize len = 0;
+
+  if (config == NULL)
+    {
+      message ("Failed to copy repository config");
```

Comment:
The error will be printed anyway, in `main`. Here, I would need to introduce a local error and propagate it then.